### DISCUSSION TITLE
CacheWriterAdapter.class

### DISCUSSION THREAD

**Discussion by manishkrishan** [ben-manes/caffeine#615] [Q&A] [answered]
Hi,

I downloaded https://repo1.maven.org/maven2/com/github/ben-manes/caffeine/caffeine/2.9.2/caffeine-2.9.2.jar and I can see `com/github/benmanes/caffeine/cache/Caffeine$CacheWriterAdapter.class` in jar but when I pick the code `https://github.com/ben-manes/caffeine/archive/refs/tags/v2.9.2.zip` and simply build using `gradlew build` with JAVA 8, I don't see this CacheWriterAdapter.class.

Just want to know how https://repo1.maven.org/maven2/com/github/ben-manes/caffeine/caffeine/2.9.2/caffeine-2.9.2.jar was built, from where I can find the jar corresponding to v2.9.2.zip source code ?

Regards,

**Comment by ben-manes** [comment, marked as answer]
The publication is performed by a [github action](https://github.com/ben-manes/caffeine/blob/v2.9.2/.github/workflows/release.yml) in response to creating the release tag. 

If you run `gradlew compileJava` then it will appear under `caffeine/build/classes/java/main/com/github/benmanes/caffeine/cache/Caffeine$CacheWriterAdapter.class`.

If you run `gradlew jar` then it will be included in `unzip -l caffeine/build/libs/caffeine-2.9.2-SNAPSHOT.jar`.

**Reply by manishkrishan** [reply]
Thankyou, I am trying it on Windows 10 and it is failing 

~~~
D:\caffeine-2.9.2>gradlew compileJava
checksum-dependency-plugin is disabled since checksumIgnore property is present
> Task :guava:compileJava FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':guava:compileJava' with following:

> Illegal/unsupported escape sequence near index 21
  D:\VA\caffeine-2.9.2\guava\build/generated-sources/.*
                       ^
~~~

any workaround or I need to try on Linux.

**Reply by ben-manes** [reply]
This was fixed on master. The errorprone path setting is a regex so it broke on Windows.
https://github.com/ben-manes/caffeine/blob/1afd872968c5a50cddb05a357e23fc21644a9a52/gradle/codeQuality.gradle#L140

The [fix commit](https://github.com/ben-manes/caffeine/commit/a7b18fa5ed56ceda5be783dd6cd267568b0334d0#diff-452bea82be696012c633c73ca1798d7b5cc028f4677ec0fc4ef073bbf6ae0a24 ) uses ` '.*/generated-sources/.*'`.

**Reply by manishkrishan** [reply]
after this change both commands `gradlew compileJava` and `gradlew  jar` done successfully. I see a `caffeine-2.9.2\build\libs\caffeine_local.jar` but it still don't have  any `com/github/benmanes/caffeine/cache/Caffeine$CacheWriterAdapter.class`

Also is there anyway to generate a pom file to see all dependencies.

**Reply by ben-manes** [reply]
The local jar is downloaded to workaround that the errorprone compiler uses caffeine, so Gradle rewrites the transitive dependency to use the project’s build. That circular behavior breaks the build.

If you run the jar task you will see a snapshot dependency and the class as described above. I did this on a macbook using the source url in your post.

**Reply by ben-manes** [reply]
You can run `gradlew caffeine:dependencies` to see the tree. Offhand I forget the task that makes the pom.

**Reply by manishkrishan** [reply]
you meant, `caffeine_local.jar` is not built from source but downloaded ? Let me try compiling this on some other OS.

**Reply by ben-manes** [reply]
Yes, `caffeine_local.jar` is a compiler dependency.

https://github.com/ben-manes/caffeine/blob/a1a5de80779780dc3844b473fe14b55180572cbf/gradle/codeQuality.gradle#L27-L45

That version uses `2.8.8` which did not have this source file, which was introduced in `2.9` with `EvictionListener` and deprecation of `CacheWriter`. When you build the jar then a `SNAPSHOT` version will be created from the project's source code and it will include the class.

**Reply by manishkrishan** [reply]
I tried to build it on Linux as well as on MacBook (version 11.5.1) with jdk1.8.0_261 using `gradlew compileJava` and `gradlew jar`, there is no compile error but same time, I don't find any `SNAPSHOT` version of jar. I only see `caffeine_local.jar` in `build/libs` folder.

**Reply by ben-manes** [reply]
You need to look at the subproject build directory, caffeine/build/libs

**Reply by manishkrishan** [reply]
thanks a lot, now I see following 2 in `caffeine/build/libs`

~~~
[user1@lnxvm libs]$ ls -l
total 1808
-rw-r--r-- 1 user1 dba 925294 Oct 25 07:20 caffeine-2.9.2.jar
-rw-r--r-- 1 user1 dba 925306 Oct 25 07:19 caffeine-2.9.2-SNAPSHOT.jar
[user1@lnxvm libs]$ pwd
/caffeine292/caffeine-2.9.2/caffeine/build/libs
~~~

What is the difference in these 2 ? I compiled with `-Prelease`. I think `caffeine-2.9.2.jar` is final release build.

**Reply by ben-manes** [reply]
Did you build the using `-Prelease`? That toggles the version string. You can use the clean task, there should be only one jar.

**Reply by manishkrishan** [reply]
yes, compiled with `-Prelease` as well based on [git action#L31](https://github.com/ben-manes/caffeine/blob/a1a5de80779780dc3844b473fe14b55180572cbf/.github/workflows/release.yml#L31).

Now looks all good, thankyou for your prompt replies.
